### ISSUE TITLE
ARK Server Crash When Using -automanagedmods In Command Line

### ISSUE THREAD

**Issue by GT500org** [Pterodactyl/Panel#257]
ARK servers need to have a copy of SteamCMD in the following folder, otherwise they will crash when using the -automanagedmods argument.
```
../Engine/Binaries/ThirdParty/SteamCMD/Linux
```
See [this](https://steamcommunity.com/app/346110/discussions/0/353915309342229013/#c353915953244644936).

**Comment by GT500org** [comment]
This is the script that is supposed to be executed before running an ARK server on Linux (you'll find it in the root of the server's data directory):
```
#!/bin/bash
# SteamCMD installer

sudo apt-get install lib32gcc1
cd Engine
cd Binaries
mkdir ThirdParty
cd ThirdParty
mkdir SteamCMD
cd SteamCMD
mkdir Linux
cd Linux
wget https://steamcdn-a.akamaihd.net/client/installer/steamcmd_linux.tar.gz
tar -xf steamcmd_linux.tar.gz
```
Obviously that won't entirely work on a CentOS/RedHat system, but you can replace the `sudo apt-get` line with something like the following, as long as it is executed with root privileges:
```
yum -y install glibc.i686 libstdc++.i686
```
If you can execute it in an interactive fashion, where the user can enter their root password when prompted in the terminal, then the following should work:
```
su - -c 'yum -y install glibc.i686 libstdc++.i686'
```

**Comment by schrej** [comment]
Which package manager to use doesn't depend on the host OS but the OS used in the container. It's probably alpine linux in this case as we try to keep images very small.
Have you tried creating the folders and downloading steamcmd to the mentioned folder manually?

**Comment by GT500org** [comment]
Yes, creating the files manually does resolve the problem, so long as the dependencies (glibc.i686 and libstdc++.i686) are installed.

I wasn't aware that your container had a virtual operating system in it. That explains the rediculous memory and CPU usage displayed in the panel, even though the system monitors all showed shootergame only using 2.9GB of RAM and no CPU time.

BTW: Did you make [these changes](http://ark.gamepedia.com/Dedicated_Server_Setup#Open_Files_Limit) in your virtual OS that you run in your container? If not, then that might explain why the ARK server doesn't work.

**Comment by schrej** [comment]
As far as i'm aware there is no higher ulimit enabled in our container. Also i think that needs to be done using [this](https://docs.docker.com/engine/reference/commandline/run/#set-ulimits-in-container---ulimit) docker flag.

**Comment by GT500org** [comment]
Unfortunately I don't have any experience with Docker.

**Comment by parkervcp** [comment]
I thought we had this already as a problem with the java thing we ran into once long ago. I also thought Fish updated that already in a previous fix as well.

**Comment by parkervcp** [comment]
@GT500org All containers are not so much a virtual OS but a chroot system where it uses the host kernel to boot inside the container. It makes all the stuff smaller.

I will also look more into the ARK container as I am interested in having it as well.

**Comment by sirmaxim** [comment]
I saw the issue with steamcmd not being in the `ThirdParty/SteamCMD/Linux/` path and took a shot at adding something to the `start.sh` based on the SteamCMDInstall.sh here: https://github.com/sirmaxim/Docker-SRCDS/tree/ark

I made a poor man's attempt to test it by dropping it strait into the `...aufs/mnt/<hash>/start.sh`. Works as intended and populates the Linux/ directory with the correct files.
I was going to make a pull request, but it still doesn't fix the CLI option `-automanagedmods` in my tests. It seems there is something else that needs to be done as well.

Normally, ARK forks a ton of steamcmd processes to update mods when this works correctly, but it appears not to fork any processes and goes strait to starting the server.

**Comment by GT500org** [comment]
When it executed the Steam installer, it ran without errors? If it did, then the game should have no trouble executing it, and the server shouldn't crash on start.

Note that regardless of whether my ARK server crashes, it never binds its ports, and thus nothing is able to communicate with it. I am assuming that this is due to the server not being able to finish starting the game for some reason, but the ARK server gives pretty much no debug information aside from a vague Steam related error that may or may not be a problem. My best guess, from what I've found around the Internet, is that the open files limit in the container needs to be increased.

**Comment by DaneEveritt** [comment]
I dont think that vague steam error is the issue since it happens with other game. Can you confirm that the startup commands between a game managed by the panel and via your own nonpanel system are the same, or if nit what the difference might be?

**Comment by GT500org** [comment]
BTW: @sirmaxim is correct that the server doesn't start any SteamCMD processes when launched via the panel (I didn't see that part of his message earlier). In fact, aside from loading about 3GB of data into RAM, it doesn't seem to _do_ anything.

**Comment by sirmaxim** [comment]
@DaneEveritt I am trying to figure the differences, but there really isn't anything that should be different about the startup command. I'll keep investigating, though. *something* is different. FWIW, I used https://gameservermanagers.com/lgsm/arkserver/ with great success to setup a server in a Debian 8 KVM. It also required one minor tweak to get the `ThirdParty/SteamCMD/Linux/` path populated, but worked perfectly otherwise. It just doesn't have a nice control panel like this.

@GT500org You have something odd going on if it's not binding to ports. You need 2 ports in your startup like this in your launch in the panel: 
`{{ARK_MAP}}?listen?ServerAdminPassword={{ARK_ADMIN_PASSWORD}}?QueryPort={{SERVER_PORT}}?Port={{UDP_PORT}}?MaxPlayers={{SERVER_MAX_PLAYERS}} -automanagedmods`
and you need to attach the other port (UDP_PORT) in the Build Configuration tab at `https://name.tld/admin/servers/view/<server_id>?tab=tab_build`

If that doesn't work for you, you did something wrong somewhere. If it is taking up ~3.5 GB of RAM, the server *is* running.

**Comment by GT500org** [comment]
@sirmaxim I don't think it's going to bind the game client port or the raw UDP socket port until it has been able to download workshop mods (it is binding the query port and the RCON port now, although I'm not sure if I verified those ports before). Since I can't debug changing the open files limit in the container, at the moment I can only assume that is the issue.

**Comment by sirmaxim** [comment]
@GT500org The open files limit would prevent the server from loading 3.5 GB of RAM. I don't think that's what it is. If it were, it would stop at < 2GB and not run at all. ARK is fat and reads a lot of files. I think it more likely there's a docker restriction of some kind preventing it from forking the steamcmd, but I don't really know docker since I avoided it for a long time having been wary of the 1.0 release not being production ready like the sales pitch said.

**Comment by DaneEveritt** [comment]
Can you use the same port for the query port as well, or do they need to be different?

**Comment by sirmaxim** [comment]
The ports must be different. One is for steam (defaults to 27015) and the other is for game clients over UDP (defaults to 7777 or 7778). Steam is required for the clients to find/connect as far as I can tell.

**Comment by GT500org** [comment]
The ports used by the server are detailed [here](http://ark.gamepedia.com/Dedicated_Server_Setup#Network).

**Comment by spannerman79** [comment]
I only just came across this issue.
Perhaps using https://github.com/FezVrasta/ark-server-tools#ark-survival-evolved-linux-server-tools within the "ARK" docker image instead would make it better and easier.

I've tried to use LGSM before and in fact it doesn't work with `-automanagedmods` as I outlined here ( https://github.com/GameServerManagers/LinuxGSM/issues/1336#issuecomment-284598650 )

**Comment by DaneEveritt** [comment]
I believe this has been fixed with the new service option scripts included in `v0.6.0-beta.3`.
